packages/sources/coinmarketcap-test/src/crypto-utils.ts

```diff
@@ -0,0 +1,188 @@
+    description: 'The symbol of the currency to convert to',
+    required: true,
+    type: 'string',
+  },
+  cid: {
+    description: 'The CMC coin ID (optional to use in place of base)',
+    required: false,
+    type: 'string',
+  },
+  slug: {
+    description: 'The CMC coin ID (optional to use in place of base)',
+    required: false,
+    type: 'string',
+  },
+} as const
+
+export interface ProviderRequestBody {
+  ids: string
+  vs_currencies: string
+  include_market_cap?: boolean
+  include_24hr_vol?: boolean
+}
+
+interface PriceInfo {
+  price: number
+  volume_24h: number
+  percent_change_1h: number
+  percent_change_24h: number
+  percent_change_7d: number
+  percent_change_30d: number
+  market_cap: number
+}
+
+export interface ProviderResponseBody {
+  data: {
+    [key: string]: {
+      id: number
+      name: string
+      symbol: string
+      slug: string
+      is_active: number
+      is_fiat: number
+      circulating_supply: number
+      total_supply: number
+      max_supply: number
+      date_added: string
+      num_market_pairs: number
+      cmc_rank: number
+      last_updated: string
+      tags: string[]
+      platform: string
+      quote: {
+        [key: string]: PriceInfo
+      }
+    }
+  }
+  status: {
+    timestamp: string
+    error_code: number
+    error_message: string
+    elapsed: number
+    credit_count: number
+  }
+  cost: number
+}
+
+export type CryptoEndpointTypes = {
+  Request: {
+    Params: CryptoRequestParams
+  }
+  Response: SingleNumberResultResponse
+  CustomSettings: typeof customSettings
+  Provider: {
+    RequestBody: ProviderRequestBody
+    ResponseBody: ProviderResponseBody
+  }
+}
+
+const groupByBaseOptions = (params: CryptoRequestParams[]) => {
+  const id: CryptoRequestParams[] = []
+  const slug: CryptoRequestParams[] = []
+  const symbol: CryptoRequestParams[] = []
+  params.forEach((param) => {
+    let optionContainer = []
+    if (param.cid) {
+      optionContainer = id
+    } else if (param.slug) {
+      optionContainer = slug
+    } else if (param.base) {
+      optionContainer = symbol
+    }
+    optionContainer.push(param)
+  })
+  return Object.entries({ id, slug, symbol })
+}
+
+export const buildBatchedRequestBody = (params: CryptoRequestParams[], endpoint: string) => {
+  //Coinmarketcap supports 3 different options for sending base params (id) - 'cid', 'slug' and 'base'. We group here to send batch requests for each such option
+  const groupedParams = groupByBaseOptions(params)
+
+  return groupedParams.map((group) => {
+    const queryName = group[0]
+    const groupedParams = group[1]
+
+    return {
+      params: groupedParams,
+      request: {
+        baseURL: endpoint,
+        url: '/cryptocurrency/quotes/latest',
+        params: {
+          [queryName]: [...new Set(groupedParams.map((p) => p.cid || p.slug || p.base))].join(','),
+          convert: [...new Set(groupedParams.map((p) => p.quote))].join(','),
```

Comment:
Based on the documentation, I think CMC limits convert to 120 symbols at a time if I'm reading the docs right. We had a limit for this in v2 also. https://github.com/smartcontractkit/external-adapters-js/blob/6e9f97e33eeaa1678d99ad0de1f5215ff45ee6fe/packages/sources/coinmarketcap/src/endpoint/crypto.ts#L14